<a href="https://colab.research.google.com/github/cokerjesutayo/machine_learning_assignment/blob/main/Assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Section 1: Data Exploration

In [ ]:
import pandas as pd

## 1. Loading the dataset

In [ ]:
df=pd.read_csv("/content/housing_dataset.csv")

In [ ]:
df

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0
...,...,...,...,...,...,...,...,...,...
17015,-122.27,37.81,40.0,880.0,451.0,582.0,380.0,0.9770,-118800.0
17016,-119.70,34.41,19.0,2086.0,575.0,1701.0,530.0,2.8042,-236100.0
17017,-115.53,34.91,12.0,807.0,199.0,246.0,102.0,2.5391,-40000.0
17018,-120.15,39.80,19.0,785.0,151.0,366.0,140.0,3.0625,-82500.0


### 1.1 Displaying the first 5 rows

In [ ]:
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


## 2. Checking for shape of the dataframe

In [ ]:
df.shape

(17020, 9)

It shows that the dataset has 17020 rows and 9 columns.

## 3. Checking for the information which shows column names, data types, and non-null counts for every column.


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17020 entries, 0 to 17019
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           17020 non-null  float64
 1   latitude            17020 non-null  float64
 2   housing_median_age  17020 non-null  float64
 3   total_rooms         17020 non-null  float64
 4   total_bedrooms      16990 non-null  float64
 5   population          16980 non-null  float64
 6   households          17020 non-null  float64
 7   median_income       16990 non-null  float64
 8   median_house_value  17020 non-null  float64
dtypes: float64(9)
memory usage: 1.2 MB


## 4.Getting the statistical summary of the DataFrame.

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
longitude,17020.0,-119.562066,2.005157,-124.3500,-121.7900,-118.49000,-118.0000,-114.3100
latitude,17020.0,35.625785,2.137446,32.5400,33.9300,34.25000,37.7200,41.9500
housing_median_age,17020.0,28.590541,12.588782,1.0000,18.0000,29.00000,37.0000,52.0000
total_rooms,17020.0,2864.361927,9385.577088,2.0000,1462.0000,2127.50000,3153.0000,379370.0000
total_bedrooms,16990.0,539.285756,421.598465,1.0000,296.0000,434.00000,648.7500,6445.0000
population,16980.0,1429.529505,1148.157282,3.0000,789.0000,1166.50000,1721.0000,35682.0000
households,17020.0,501.103408,384.400904,1.0000,282.0000,409.00000,605.2500,6082.0000
median_income,16990.0,3.883400,1.908640,0.4999,2.5659,3.54425,4.7676,15.0001
median_house_value,17020.0,205747.585488,118642.283445,-500001.0000,118800.0000,179800.00000,264700.0000,500001.0000


I noticed that the data are in different scales because  total_rooms, total_bedrooms, population, households, and median_house_value have very large values, while others like median_income, housing_median_age, latitude, and longitude have  smaller values

## 5. Checking missing values




In [ ]:
df.isnull().sum()

,0
longitude,0
latitude,0
housing_median_age,0
total_rooms,0
total_bedrooms,30
population,40
households,0
median_income,30
median_house_value,0


total_bedrooms and median_income has 30 missing values while population has 40 missing values.

## 6. Checking for duplicates

In [ ]:
df.duplicated().sum()

np.int64(20)

# Section 2: Data Cleaning

## 7. Handling missing values

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
longitude,17020.0,-119.562066,2.005157,-124.3500,-121.7900,-118.49000,-118.0000,-114.3100
latitude,17020.0,35.625785,2.137446,32.5400,33.9300,34.25000,37.7200,41.9500
housing_median_age,17020.0,28.590541,12.588782,1.0000,18.0000,29.00000,37.0000,52.0000
total_rooms,17020.0,2864.361927,9385.577088,2.0000,1462.0000,2127.50000,3153.0000,379370.0000
total_bedrooms,16990.0,539.285756,421.598465,1.0000,296.0000,434.00000,648.7500,6445.0000
population,16980.0,1429.529505,1148.157282,3.0000,789.0000,1166.50000,1721.0000,35682.0000
households,17020.0,501.103408,384.400904,1.0000,282.0000,409.00000,605.2500,6082.0000
median_income,16990.0,3.883400,1.908640,0.4999,2.5659,3.54425,4.7676,15.0001
median_house_value,17020.0,205747.585488,118642.283445,-500001.0000,118800.0000,179800.00000,264700.0000,500001.0000


#### longitude

In [ ]:
longitude_mean= df["longitude"].mean()
print(longitude_mean)

-119.56206580493536


In [ ]:
df["longitude"].isnull().sum()

np.int64(0)

I choose to use mean because the gap between those numbers are not too much in the longitude and no outliers . No missing values for this one

#### latitude

In [ ]:
latitude_mean= df["latitude"].mean()
print(latitude_mean)

35.62578495887191


In [ ]:
df["latitude"].isnull().sum()

np.int64(0)

I choose to use mean because the gap between those numbers are not too much in the latitude and no outliers . No missing values for this one

#### housing median age

In [ ]:
housing_median_age=df["housing_median_age"].median()
print(housing_median_age)

29.0


In [ ]:
df["housing_median_age"].isnull().sum()

np.int64(0)

I choose to use median because the gap between those numbers are too much in the housing median age and that is an outliers . No missing values for this one

#### total rooms

In [ ]:
total_rooms_median=df["total_rooms"].median()
print(total_rooms_median)

2127.5


In [ ]:
df["total_rooms"].isnull().sum()

np.int64(0)

I choose to use median because the gap between those numbers are too much in the total rooms and that is an outliers . No missing values for this one

#### total bedrooms

In [ ]:
total_bedrooms_median=df["total_bedrooms"].median()
print(total_bedrooms_median)

434.0


In [ ]:
df['total_bedrooms']=df["total_bedrooms"].fillna(total_bedrooms_median)

I choose to use median because the gap between those numbers are too much in the total bedrooms and that is an outliers . There is missing value for this one so we use fillna to fill the median number into it

In [ ]:
df['total_bedrooms'].isnull().sum()

np.int64(0)

#### population

In [ ]:
population_median=df["population"].median()
print(population_median)

1166.5


In [ ]:
df["population"]=df["population"].fillna(population_median)

I choose to use median because the gap between those numbers are too much in the population median and that is an outliers . There is missing value for this one so we use fillna to fill the median number into it

In [ ]:
df["population"].isnull().sum()

np.int64(0)

#### household


In [ ]:
households_median= df["households"].median()
print(households_median)

409.0


In [ ]:
df["households"].isnull().sum()

np.int64(0)

I choose to use median because the gap between those numbers are too much in the household and that is an outliers . No missing values for this one

#### median income

In [ ]:
income_median=df["median_income"].median()
print(income_median)

3.54425


In [ ]:
df["median_income"]=df["median_income"].fillna(income_median)

I choose to use median because the gap between those numbers are too much in the median income and that is an outliers . There is missing value for this one so we use fillna to fill the median number into it

In [ ]:
df["median_income"].isnull().sum()

np.int64(0)

#### median house value


In [ ]:
house_value_median= df['median_house_value'].median()
print(house_value_median)

179800.0


In [ ]:
df['median_house_value'].isnull().sum()

np.int64(0)

I choose to use median because the gap between those numbers are too much in the median house value and that is an outliers . No missing values for this one

## 8. Handling duplicates

In [ ]:
df.shape

(17020, 9)

In [ ]:
df=df.drop_duplicates()

In [ ]:
df.shape

(17000, 9)

Keeping duplicate rows in a dataset is a problem because they repeat the same information which can bias the model, make it memorize data instead of learning patterns and reduce its performance wheen training new data.


# Section 3. Feature Scaling

Feature scaling is the process of bringing all data values to the same range so no feature is too big or too small.

Why is it necessary?
Feature scaling makes all values similar in size so no feature dominates because without it large values can mislead the model.


The model would focus too much on features with large values and ignore the smaller ones leading to poor and inaccurate results.

## 10. Min-Max

In [ ]:
cols_to_scale=["housing_median_age","total_rooms","median_income","median_house_value"]

In [ ]:
for col in cols_to_scale:
  col_min=df[col].min() #find the min value in the column
  col_max=df[col].max() #find the max value in the column
  df[col+"_scaled"]=(df[col]- col_min)/ (col_max-col_min) #Apply the formula to every row in the column

/tmp/ipykernel_7651/1922946507.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col+"_scaled"]=(df[col]- col_min)/ (col_max-col_min) #Apply the formula to every row in the column
/tmp/ipykernel_7651/1922946507.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col+"_scaled"]=(df[col]- col_min)/ (col_max-col_min) #Apply the formula to every row in the column
/tmp/ipykernel_7651/1922946507.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .lo

In [ ]:
#check that every scaled column is now between 0 and 1
for col in cols_to_scale:
  scaled_col=col+"_scaled"
  print(f"{scaled_col} : min {df[scaled_col].min():.4f} max {df[scaled_col].max():.4f}")

housing_median_age_scaled : min 0.0000 max 1.0000
total_rooms_scaled : min 0.0000 max 1.0000
median_income_scaled : min 0.0000 max 1.0000
median_house_value_scaled : min 0.0000 max 1.0000


In [ ]:
print(df[[col, col+"_scaled"]].head(10))

   median_house_value  median_house_value_scaled
0             66900.0                     0.5669
1             80100.0                     0.5801
2             85700.0                     0.5857
3             73400.0                     0.5734
4             65500.0                     0.5655
5             74000.0                     0.5740
6             82400.0                     0.5824
7             48500.0                     0.5485
8             58400.0                     0.5584
9             48100.0                     0.5481


The scaled values are just the original values made smaller so they all fit between 0 and 1 so  when the original number increases, the scaled one also increases only the size changes